store/postgres/src/mapping.rs

```diff
@@ -0,0 +1,1081 @@
+        Ok(ClampRangeQuery::new(&self.schema, table, key, None, block).execute(conn)?)
+    }
+
+    pub fn revert_block(
+        &self,
+        conn: &PgConnection,
+        block: BlockNumber,
+    ) -> Result<(StoreEvent, i32), StoreError> {
+        let mut changes: Vec<EntityChange> = Vec::new();
+        let mut count: i32 = 0;
+
+        for table in self.tables.values() {
+            // Remove all versions whose entire block range lies beyond
+            // `block`
+            let removed = RevertRemoveQuery::new(&self.schema, table, block)
+                .get_results(conn)?
+                .into_iter()
+                .map(|data| data.id)
+                .collect::<HashSet<_>>();
+            // Make the versions current that existed at `block - 1` but that
+            // are not current yet. Those are the ones that were updated or
+            // deleted at `block`
+            let unclamped = RevertClampQuery::new(&self.schema, table, block - 1)
+                .get_results(conn)?
+                .into_iter()
+                .map(|data| data.id)
+                .collect::<HashSet<_>>();
+            // Adjust the entity count; we can tell which operation was
+            // initially performed by
+            //   id in (unset - unclamped)  => insert (we now deleted)
+            //   id in (unset && unclamped) => update (we reversed the update)
+            //   id in (unclamped - unset)  => delete (we now inserted)
+            let deleted = removed.difference(&unclamped).count() as i32;
+            let inserted = unclamped.difference(&removed).count() as i32;
+            count += inserted - deleted;
+            // EntityChange for versions we just deleted
+            let deleted = removed
+                .into_iter()
+                .filter(|id| !unclamped.contains(id))
+                .map(|id| EntityChange {
+                    subgraph_id: self.subgraph.clone(),
+                    entity_type: table.object.clone(),
+                    entity_id: id,
+                    operation: EntityChangeOperation::Removed,
+                });
+            changes.extend(deleted);
+            // EntityChange for versions that we just updated or inserted
+            let set = unclamped.into_iter().map(|id| EntityChange {
+                subgraph_id: self.subgraph.clone(),
+                entity_type: table.object.clone(),
+                entity_id: id,
+                operation: EntityChangeOperation::Set,
+            });
+            changes.extend(set);
+        }
+        Ok((StoreEvent::new(changes), count))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Reference {
+    pub table: SqlName,
+    pub column: SqlName,
+}
+
+impl Reference {
+    fn to_column(table: SqlName, column: &Column) -> Reference {
+        let column = match &column.derived {
+            Some(col) => col,
+            None => PRIMARY_KEY_COLUMN,
+        }
+        .into();
```

Comment:
I find it strange that a derived field is has a reference. It seems we don't yet generate sql queries for derived columns so the `col` value here would never be used right? This is confusing to me.

Reply:
Commit efc1882 makes it so that we completely ignore derived fields in the relational schema.